### PR TITLE
build: GCC-12 build improvements

### PR DESCRIPTION
**Pull request by fanquake**
2 changes to better support building with GCC 12, which out of the box, is currently broken if you want to build using depends.
Prevent `-Warray-bounds` errors when building libxkbcommon. i.e:
```bash
src/xkbcomp/ast-build.c:82:27: error: array subscript 'ExprDef[0]' is partly outside array bounds of 'unsigned char[32]' [-Werror=array-bounds]
   82 |     expr->expr.value_type = type;
      |     ~~~~~~~~~~~~~~~~~~~~~~^~~~~~
src/xkbcomp/ast-build.c:75:21: note: object of size 32 allocated by 'malloc'
   75 |     ExprDef *expr = malloc(size);
      |                     ^~~~~~~~~~~~
```

It might be the case that these would be fixed by updating the
package, but that would also require installing new build tools (meson),
as well as potentially more dependencies (wayland), and it'd need
testing with Qt. For now, just turn the errors into wanrings.

Define `BOOST_NO_CXX98_FUNCTION_BASE` to prevent GCC warning about the use of `std::unary_function`. i.e:
```bash
/bitcoin/depends/aarch64-unknown-linux-gnu/include/boost/container_hash/hash.hpp:131:33:
warning: 'template<class _Arg, class _Result> struct std::unary_function' is deprecated [-Wdeprecated-declarations]
  131 |         struct hash_base : std::unary_function<T, std::size_t> {};
      |                                 ^~~~~~~~~~~~~~
In file included from /usr/include/c++/12/bits/unique_ptr.h:37,
                 from /usr/include/c++/12/memory:76,
                 from ./init.h:10,
                 from init.cpp:10:
/usr/include/c++/12/bits/stl_function.h:117:12: note: declared here
  117 |     struct unary_function
```

Boost `container_hash` (included via functional -> multi_index) uses
[`std::unary_function`, which was deprecated in C++11](https://en.cppreference.com/w/cpp/utility/functional/unary_function), and "removed" in
C++17. It's use causes warnings with newer compilers, i.e GCC 12.1.

Use the MACRO outlined in https://github.com/boostorg/container_hash/issues/22, and added to Boost Config for GCC 12 in https://github.com/boostorg/config/pull/430, to prevent it's use.

[BOOST_NO_CXX98_FUNCTION_BASE](https://www.boost.org/doc/libs/master/libs/config/doc/html/boost_config/boost_macro_reference.html):
> The standard library no longer supports std::unary_function and std::binary_function.
> They were deprecated in C++11 and is removed from C++14.

Guix Build (x86_64):
```bash

```

Guix Build (arm64):
```bash

```

### CODE REVIEW

**Comment by jarolrod** [comment]
GUIX hashes

x86:
```
$ find guix-build-$(git rev-parse --short=12 HEAD)/output/ -type f -print0 | env LC_ALL=C sort -z | xargs -r0 sha256sum

648beb27bb60aad6b88c0865293738665d55421cb863a17ced246ea4fad5f18c  guix-build-f1df4dd2878a/output/aarch64-linux-gnu/SHA256SUMS.part
0901076bb0055522b0d733b7efc51afedec82b8ba4d3c8fb5cfb80dc8e5d58d6  guix-build-f1df4dd2878a/output/aarch64-linux-gnu/bitcoin-f1df4dd2878a-aarch64-linux-gnu-debug.tar.gz
e743b9acd4e6e3263328896409e37b0e702637294b435f09cde899f6d123bcb6  guix-build-f1df4dd2878a/output/aarch64-linux-gnu/bitcoin-f1df4dd2878a-aarch64-linux-gnu.tar.gz
cf37822a0a23af22a22d3b3447dc8540cae604df68881625948e151ff6d978b9  guix-build-f1df4dd2878a/output/arm-linux-gnueabihf/SHA256SUMS.part
a2373bc8e084697bbec1e08b40dc3d85d08dd922cea34b99739b79ce35d07ee2  guix-build-f1df4dd2878a/output/arm-linux-gnueabihf/bitcoin-f1df4dd2878a-arm-linux-gnueabihf-debug.tar.gz
2092364a132b6028f0c5c3a8b8a716fbd592089245ad74c1b8261da88dbd0e88  guix-build-f1df4dd2878a/output/arm-linux-gnueabihf/bitcoin-f1df4dd2878a-arm-linux-gnueabihf.tar.gz
6568fcfd0babeace56362e6921490ab50a6a84ee28aa248606df27d3cd862383  guix-build-f1df4dd2878a/output/arm64-apple-darwin/SHA256SUMS.part
fe98876ec35eded87e464ced7c90f3dcfc7961e23b03ef0e3e6e9c4ecf9b35bf  guix-build-f1df4dd2878a/output/arm64-apple-darwin/bitcoin-f1df4dd2878a-arm64-apple-darwin-unsigned.dmg
9cf0fb7c2b430c877e828aa4c942e3ffbee2f36d6acbf783dde9735e672ba6ce  guix-build-f1df4dd2878a/output/arm64-apple-darwin/bitcoin-f1df4dd2878a-arm64-apple-darwin-unsigned.tar.gz
f18ebae8ac75518dd21555c69bfed16a6a0a5b2590afaf96e6a6ffff9bc34d7f  guix-build-f1df4dd2878a/output/arm64-apple-darwin/bitcoin-f1df4dd2878a-arm64-apple-darwin.tar.gz
d939b5050fe48bade289a6209a147e9b2685d8784c7d328321462aa573724440  guix-build-f1df4dd2878a/output/dist-archive/bitcoin-f1df4dd2878a.tar.gz
3ec352d2321a358d1009b4ce0635782989de509b0dc82e9254e222b0a9220bf9  guix-build-f1df4dd2878a/output/powerpc64-linux-gnu/SHA256SUMS.part
d1fc1df89ba92debdc7ca1f865edbbf6d86365a2ffbf426705f708c494defe1e  guix-build-f1df4dd2878a/output/powerpc64-linux-gnu/bitcoin-f1df4dd2878a-powerpc64-linux-gnu-debug.tar.gz
ec15d97095b46b85e68dafce1f80639cccb6baa3cb141dcda1d65a646b135d57  guix-build-f1df4dd2878a/output/powerpc64-linux-gnu/bitcoin-f1df4dd2878a-powerpc64-linux-gnu.tar.gz
0b4dd2a8beb64513b14e5ce14004d25b3bc59bdff7401594f8278e5ebf682871  guix-build-f1df4dd2878a/output/powerpc64le-linux-gnu/SHA256SUMS.part
d72d5fd1aa3626cae8599cc51d0a4a19612edd4ae13c49dce9f9334a9a7c9037  guix-build-f1df4dd2878a/output/powerpc64le-linux-gnu/bitcoin-f1df4dd2878a-powerpc64le-linux-gnu-debug.tar.gz
57ed26b03b28cc00e7ae49fac049e43bb39fa57b33bbfaae5a1d37a515287fe7  guix-build-f1df4dd2878a/output/powerpc64le-linux-gnu/bitcoin-f1df4dd2878a-powerpc64le-linux-gnu.tar.gz
15847353a751ccc323040e2c05033860a2c4a77381b1e87fe3c7a158fd1c0ffc  guix-build-f1df4dd2878a/output/riscv64-linux-gnu/SHA256SUMS.part
8a8620d52cc639da368adb79c57337cbb4ec0e17233f8fdbd18760f4700a364b  guix-build-f1df4dd2878a/output/riscv64-linux-gnu/bitcoin-f1df4dd2878a-riscv64-linux-gnu-debug.tar.gz
3d647c1e21d590b73758baf8f2a87adf0230d6ced9ac1310f8f3db3d50095789  guix-build-f1df4dd2878a/output/riscv64-linux-gnu/bitcoin-f1df4dd2878a-riscv64-linux-gnu.tar.gz
5df51ce85386708b33c8004394e6a120eaee62c80bc6f928f1addc1d87cb8f90  guix-build-f1df4dd2878a/output/x86_64-apple-darwin/SHA256SUMS.part
53ee3557b921c2c218ecfca89896eb0e547f652dea26a2630a0b13203f66f73d  guix-build-f1df4dd2878a/output/x86_64-apple-darwin/bitcoin-f1df4dd2878a-x86_64-apple-darwin-unsigned.dmg
be21aec933ce47ca5a8fc0838e211aaf8f11234e3b5f29a17a1589914c764121  guix-build-f1df4dd2878a/output/x86_64-apple-darwin/bitcoin-f1df4dd2878a-x86_64-apple-darwin-unsigned.tar.gz
37e3e633584e4a7060cda581a6097df769b9d2195a28654bc07e2bf1cb4576f5  guix-build-f1df4dd2878a/output/x86_64-apple-darwin/bitcoin-f1df4dd2878a-x86_64-apple-darwin.tar.gz
d2f0835c7fa07e37f551035db1300c5b631e17d28c0bc88146a053bc12a49ec7  guix-build-f1df4dd2878a/output/x86_64-linux-gnu/SHA256SUMS.part
d99d9e7bae361ed981ffad52c7c42c1e7952d9e237f27f452981c32f379c64bb  guix-build-f1df4dd2878a/output/x86_64-linux-gnu/bitcoin-f1df4dd2878a-x86_64-linux-gnu-debug.tar.gz
df36401321423afab0bd7a9bb53f2f6f8368ff3bdb80209a192512a6673106c7  guix-build-f1df4dd2878a/output/x86_64-linux-gnu/bitcoin-f1df4dd2878a-x86_64-linux-gnu.tar.gz
b41a34bce877a47eacf57eb04ca628f9310a4bd3756a7054b33c3ffa6f9ac52b  guix-build-f1df4dd2878a/output/x86_64-w64-mingw32/SHA256SUMS.part
6c1ca19f54943e4a246693698319c6b89df3a7ca7125d26ce7b73d39bf5e052f  guix-build-f1df4dd2878a/output/x86_64-w64-mingw32/bitcoin-f1df4dd2878a-win64-debug.zip
ea952cb4db8776e0b9250aed8653e309ea8824b4763853a86ef41d2912a3b99d  guix-build-f1df4dd2878a/output/x86_64-w64-mingw32/bitcoin-f1df4dd2878a-win64-setup-unsigned.exe
b5f92fc5d319beacb8b5de4af8f9ed1da5e2465246f8f1f1496378cf1aa0a17d  guix-build-f1df4dd2878a/output/x86_64-w64-mingw32/bitcoin-f1df4dd2878a-win64-unsigned.tar.gz
b45750ff21a0f88024ca34590022f71612ca369b5b98e503b3ac10231cd25c67  guix-build-f1df4dd2878a/output/x86_64-w64-mingw32/bitcoin-f1df4dd2878a-win64.zip
```

arm64:
```

```

File: depends/packages/libxkbcommon.mk [hunk 1]
```diff
@@ -8,6 +8,7 @@ $(package)_dependencies=libxcb
 define $(package)_set_vars
 $(package)_config_opts = --enable-option-checking --disable-dependency-tracking
 $(package)_config_opts += --disable-static --disable-docs
+$(package)_cflags += -Wno-error=array-bounds
```
[the remark this answers is not in the document]
Comment (fanquake) [reply depends/packages/libxkbcommon.mk:15]: I've added a comment that hopefully explains both.

**Comment by fanquake** [comment]
I've also played with updating the libxkbcommon package, which requires changing build systems (meson/ninja), there's a work in progress branch here, https://github.com/fanquake/bitcoin/commit/adb123d84b33f1eadfa416d28b2f3b093678b10e, but it doesn't yet work properly. If someone wanted to play with that, they could, but I also don't think we should block building depends with GCC 12 on updating libxkbcommon.

**Comment by laanwj** [comment]
This is fine imo. I mean we only use libxkbcommon to link against, it doesn't really matter what the version is (older is better, probably, from a compat point of view).

**Pull request merged**